### PR TITLE
v3.0, dev: add dm benchmark report

### PR DESCRIPTION
**Pull request by amyangfei**
### What is changed, added or deleted? 

DM benchmark report, Chinese version

### What is the related PR or file link(s)? 

https://github.com/pingcap/docs/pull/1521

### Which version does your change affect? 

v3.0, dev

### CODE REVIEW

File: dev/benchmark/dm-v1.0-ga.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
+---
+title: DM 1.0-GA 性能测试报告
+category: benchmark
+---
+
+# DM 1.0-GA 性能测试报告
+
+本性能报告会记录测试目的、测试环境、测试场景和测试结果。
```
Comment (GregoryIan) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
本性能报告包含 DM 1.0 GA 版本性能测试的目的、环境、场景和结果。
```

File: dev/benchmark/dm-v1.0-ga.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+```bash
+sysbench --test=oltp_insert --tables=4 --mysql-host=172.16.4.40 --mysql-port=3306 --mysql-user=root --mysql-db=dm_benchmark --db-driver=mysql --table-size=50000000 prepare
+```
+
+#### 全量导入性能测试结果
+
+| 测试项             | 导出线程数  | mydumper extra-args 参数        | 导出速度 (MB/s)   |
+| :----------------: | :---------: | :-----------------------------: | :---------------: |
+| 开启单表并发导出   | 32          | "-r 320000 --regex '^sbtest.*'" | 191.03            |
```
Comment (GregoryIan) [inline dev/benchmark/dm-v1.0-ga.md]: 要不要说明一下怎么开启 “单表并发导出”？ 貌似 mydumper 文档也没介绍

File: dev/benchmark/dm-v1.0-ga.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+| 不开启单表并发导出 | 32          | "--regex '^sbtest.*'"           | 72.22             |
```
Comment (GregoryIan) [inline dev/benchmark/dm-v1.0-ga.md]: 这样子的话导出线程数改成 4， 实际上也只用了 4

File: dev/benchmark/dm-v1.0-ga.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+sysbench --test=oltp_insert --tables=4 --num-threads=32 --mysql-host=172.17.4.40 --mysql-port=3306 --mysql-user=root --mysql-db=dm_benchmark --db-driver=mysql --report-interval=10 --time=1800 run
+```
+
+该性能测试中同步任务 `sync` 处理单元 `worker-count` 设置为 32，`batch` 大小设置为 100。
+
+| 组件                       | qps                                                          | tps                                                             | 95% 延迟                     |
+| :------------------------: | :----------------------------------------------------------: | :-------------------------------------------------------------: | :--------------------------: |
+| MySQL                      | 42.79k                                                       | 42.79k                                                          | 1.18ms                       |
+| DM relay log unit          | -                                                            | 11.3MB/s                                                        | 45us (binlog 读延迟)         |
+| DM binlog replication unit | 22.97k (binlog event 接收qps，不包括忽略的 event)            | 45.91k (DM 每秒分发的内部 job 数量，通过 query-status 查询获得) | 20ms (事务执行时间)          |
```
Comment (GregoryIan) [inline dev/benchmark/dm-v1.0-ga.md]: 45.91k 这是需要除以 2 吗？

File: dev/benchmark/dm-v1.0-ga.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+category: benchmark
+---
+
+# DM 1.0-GA 性能测试报告
+
+本性能报告会记录测试目的、测试环境、测试场景和测试结果。
+
+## 测试目的
+
+该性能测试用于评估使用 DM 进行全量数据导入和增量数据同步的性能上限，并根据测试结果提供 DM 同步任务的参考配置。
```
Comment (GregoryIan) [inline dev/benchmark/dm-v1.0-ga.md]: 可以说明测试结果是单台 dm-worker

**Comment by GregoryIan** [comment]
Rest LGTM

File: dev/benchmark/dm-v1.0-ga.md [hunk 6]
```diff
@@ -98,10 +98,12 @@ sysbench --test=oltp_insert --tables=4 --mysql-host=172.16.4.40 --mysql-port=330
 
 #### 全量导入性能测试结果
 
-| 测试项             | 导出线程数  | mydumper extra-args 参数        | 导出速度 (MB/s)   |
-| :----------------: | :---------: | :-----------------------------: | :---------------: |
-| 开启单表并发导出   | 32          | "-r 320000 --regex '^sbtest.*'" | 191.03            |
-| 不开启单表并发导出 | 32          | "--regex '^sbtest.*'"           | 72.22             |
+在 `mydumper` 中使用 `--rows` 选项可以开启单表多线程并发导出（当前 `mydumper` 在 MySQL 的单表并发导出仅支持存在主键或唯一索引，并且该列是 INT 类型），以下的第一项测试用于验证开启单表并发导出可以提高数据导出性能。
```
Comment (GregoryIan) [inline dev/benchmark/dm-v1.0-ga.md]: 也支持general index

Comment (amyangfei) [reply dev/benchmark/dm-v1.0-ga.md]: https://github.com/pingcap/mydumper/blob/master/mydumper.c#L1946
这里 `conf->use_any_index` 没有设置，总是 false，不会使用普通索引

Comment (amyangfei) [reply dev/benchmark/dm-v1.0-ga.md]: `use_any_index` is initialized at https://github.com/pingcap/mydumper/blob/master/mydumper.c#L1201

File: dev/benchmark/dm-v1.0-ga.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
+---
+title: DM 1.0-GA 性能测试报告
+category: benchmark
+---
+
+# DM 1.0-GA 性能测试报告
+
+本性能报告会记录测试目的、测试环境、测试场景和测试结果。
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
本报告记录了对 1.0-GA 版本的 DM 进行性能测试的目的、环境、场景和结果。
```

File: dev/benchmark/dm-v1.0-ga.md [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+
+## 测试目的
+
+该性能测试用于评估使用 DM 进行全量数据导入和增量数据同步的性能上限，并根据测试结果提供 DM 同步任务的参考配置。
+
+## 测试环境
+
+### 测试机器信息
+
+系统信息:
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
系统信息：
```

File: dev/benchmark/dm-v1.0-ga.md [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+| 机器 IP     | 操作系统                      | 内核版本                  | 文件系统类型     |
+| :---------: | :---------------------------: | :-----------------------: | :--------------: |
+| 172.16.4.39 | CentOS Linux release 7.6.1810 | 3.10.0-957.1.3.el7.x86_64 | ext4             |
+| 172.16.4.40 | CentOS Linux release 7.6.1810 | 3.10.0-957.1.3.el7.x86_64 | ext4             |
+| 172.16.4.41 | CentOS Linux release 7.6.1810 | 3.10.0-957.1.3.el7.x86_64 | ext4             |
+| 172.16.4.42 | CentOS Linux release 7.6.1810 | 3.10.0-957.1.3.el7.x86_64 | ext4             |
+| 172.16.4.43 | CentOS Linux release 7.6.1810 | 3.10.0-957.1.3.el7.x86_64 | ext4             |
+| 172.16.4.44 | CentOS Linux release 7.6.1810 | 3.10.0-957.1.3.el7.x86_64 | ext4             |
+
+硬件信息:
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
硬件信息：
```

File: dev/benchmark/dm-v1.0-ga.md [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+| 类别         | 指标                                                |
+| :----------: | :-------------------------------------------------: |
+| CPU          | 40 vCPUs, Intel(R) Xeon(R) CPU E5-2630 v4 @ 2.20GHz |
+| 内存         | 192GB, 12 条 16GB DIMM DDR4 2133 MHz                |
+| 磁盘         | Intel DC P4510 4TB NVMe PCIe 3.0                    |
+| 网卡         | 万兆网卡                                            |
+
+其他:
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
其他：
```

File: dev/benchmark/dm-v1.0-ga.md [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+
+* 服务器间网络延迟: rtt min/avg/max/mdev = 0.074/0.088/0.121/0.019 ms
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
* 服务器间网络延迟：rtt min/avg/max/mdev = 0.074/0.088/0.121/0.019 ms
```

File: dev/benchmark/dm-v1.0-ga.md [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+
+#### 在 load 处理单元使用不同 pool size 的性能测试对比
+
+该测试中全量导入的数据量为 3.78GB，使用以下 `sysbench` 命令生成：
+
+```
+sysbench --test=oltp_insert --tables=4 --mysql-host=172.16.4.40 --mysql-port=3306 --mysql-user=root --mysql-db=dm_benchmark --db-driver=mysql --table-size=5000000 prepare
+```
+
+| load 处理单元 pool size | 事务执行时间 (s) | 时间 (s) | 导入速度(MB/s) | TiDB 99 duration (s) |
```
Comment (csuzhangxc) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
| load 处理单元 pool size | 事务执行时间 (s) | 导入时间 (s) | 导入速度(MB/s) | TiDB 99 duration (s) |
```

File: dev/benchmark/dm-v1.0-ga.md [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+
+### 同步数据流
+
+MySQL1 (172.16.4.40) -> DM-worker1 (172.16.4.39) -> TiDB (172.16.4.41)
+
+### 公共配置信息
+
+#### 同步数据表结构
+
+```sql
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
{{< copyable "sql" >}}

```sql
```
ref: [[ICM-95] Update code block format to better support copy - PingCAP JIRA](https://internal.pingcap.net/jira/browse/ICM-95)

File: dev/benchmark/dm-v1.0-ga.md [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+  `c` char(120) CHARSET utf8mb4 COLLATE utf8mb4_bin NOT NULL DEFAULT '',
+  `pad` char(60) CHARSET utf8mb4 COLLATE utf8mb4_bin NOT NULL DEFAULT '',
+  PRIMARY KEY (`id`),
+  KEY `k_1` (`k`)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
+```
+
+#### 数据库配置
+
+使用 `TiDB-ansible` 部署 TiDB 测试集群，所有配置使用 `TiDB-ansible` 提供的默认配置。
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
使用 TiDB Ansible 部署 TiDB 测试集群，所有配置使用 TiDB Ansible 提供的默认配置。
```
Ref: [TiDB 周边工具输出名称确认 - Google 文档](https://docs.google.com/document/d/1XTjxZnpYptVFGiaT41XbvzEhFnK0Gl6bGu_WwXeFvNA/edit?ts=5c9b0829)

File: dev/benchmark/dm-v1.0-ga.md [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+
+#### 测试过程
+
+- 部署测试环境
+- 使用 `sysbench` 在上游创建测试表，并生成全量导入的测试数据
+- 在 `full` 模式下启动 DM 同步任务
+
+`sysbench` 生成数据的命令如下所示：
+
+```bash
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
{{< copyable "shell-regular" >}}

```bash
```

File: dev/benchmark/dm-v1.0-ga.md [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+
+| 测试项    | 事务执行延迟 (s) | 每条插入语句包含的行数 | 导入数据量 (GB) | 导入时间 (s) | 导入速度 (MB/s) |
+| :-------: | :--------------: | :--------------------: | :-------------: | :----------: | :-------------: |
+| load data | 1.737            | 4878                   | 38.14           | 2346.9       | 16.64           |
+
+#### 在 load 处理单元使用不同 pool size 的性能测试对比
+
+该测试中全量导入的数据量为 3.78GB，使用以下 `sysbench` 命令生成：
+
+```
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
{{< copyable "shell-regular" >}}

```bash
```

File: dev/benchmark/dm-v1.0-ga.md (hunk 12 above)
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
| load 处理单元 pool size | 事务执行时间 (s) | 时间 (s) | 导入速度 (MB/s) | TiDB 99 duration (s) |
```

File: dev/benchmark/dm-v1.0-ga.md [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+| 2                       | 0.250            | 425.9    | 9.1            | 0.23                 |
+| 4                       | 0.523            | 360.1    | 10.7           | 0.41                 |
+| 8                       | 0.986            | 267.0    | 14.5           | 0.93                 |
+| 16                      | 2.022            | 265.9    | 14.5           | 2.68                 |
+| 32                      | 3.778            | 262.3    | 14.7           | 6.39                 |
+| 64                      | 7.452            | 281.9    | 13.7           | 8.00                 |
+
+#### 导入数据时每条插入语句包含行数不同的情况下的性能测试对比
+
+该测试中全量导入的数据量为 3.78GB，load 处理单元 `pool-size` 大小为32. 插入语句包含行数通过 mydumper 的 `--statement-size` 来控制。
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
该测试中全量导入的数据量为 3.78 GB，load 处理单元 `pool-size` 大小为 32。插入语句包含行数通过 mydumper 的 `--statement-size` 来控制。
```

File: dev/benchmark/dm-v1.0-ga.md [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+| 开启单表并发导出   | 32          | "--rows 320000 --regex '^sbtest.*'" | 191.03            |
+| 不开启单表并发导出 |  4          | "--regex '^sbtest.*'"               | 72.22             |
+
+| 测试项    | 事务执行延迟 (s) | 每条插入语句包含的行数 | 导入数据量 (GB) | 导入时间 (s) | 导入速度 (MB/s) |
+| :-------: | :--------------: | :--------------------: | :-------------: | :----------: | :-------------: |
+| load data | 1.737            | 4878                   | 38.14           | 2346.9       | 16.64           |
+
+#### 在 load 处理单元使用不同 pool size 的性能测试对比
+
+该测试中全量导入的数据量为 3.78GB，使用以下 `sysbench` 命令生成：
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
该测试中全量导入的数据量为 3.78 GB，使用以下 `sysbench` 命令生成：
```

File: dev/benchmark/dm-v1.0-ga.md [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+|            310             |  -s 62500 -r 320000       |   0.413           |  322.6       |     12.0        |         0.49         |
+
+### 增量同步性能测试用例
+
+#### 测试过程
+
+- 部署测试环境
+- 使用 `sysbench` 在上游创建测试表，并生成全量导入的测试数据
+- 在 `all` 模式下启动 DM 同步任务，等待同步任务进入 `sync` 同步阶段
+- 使用 `sysbench` 在上游持续生成增量数据，通过 `query-status` 观测 DM 的同步状态，通过 grarana 观测 DM 和 TiDB 的监控指标。
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
- 使用 `sysbench` 在上游持续生成增量数据，通过 `query-status` 命令观测 DM 的同步状态，通过 Grafana 观测 DM 和 TiDB 的监控指标。
```

File: dev/benchmark/dm-v1.0-ga.md [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+
+#### 增量同步性能测试结果
+
+上游 `sysbench` 生成增量数据命令
+
+```
```
Comment (yikeke) [inline dev/benchmark/dm-v1.0-ga.md]:
```suggestion
{{< copyable "shell-regular" >}}

```bash
```

**Pull request merged**